File: LeetCode/Explore/TopInterview/Amazon/ArrayString/20_valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        mp = {')':'(', '}':'{', ']':'['}
        sp = list()
        
        for i, c in enumerate(s):
          if c not in mp.keys():
            sp.append(c)
          else:
            if len(sp) == 0 or sp[-1] != mp[c]:
              return False
            sp.pop(-1)
            
        return len(sp) == 0

File: LeetCode/Explore/TopInterview/Amazon/ArrayString/20_valid_parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
      unordered_map<char, char> mp = {{')','('}, {']','['}, {'}','{'}};
      stack<char> sp; 
      for (auto &c : s) {
        if (!mp.count(c)) {
          sp.push(c);
        }
        else {
          if (sp.empty() || sp.top() != mp[c]) {
            return false;
          }
          sp.pop();
        }
      }
      return sp.empty();
    }
};
